### PR TITLE
Prototype for safety rules counters

### PR DESCRIPTION
**Pull request by ankushagarwal**
## Summary

`safety-rules/src/counters.rs` defines counters tracked by safety rules

`consensus/safety-rules/src/simple_push_metrics.rs` is a general push metrics client which pushes metrics to a prometheus pushgateway

This is a very quick prototype of one way on how we can have counters in safety rules servers. Counters are pushed to Prometheus every 10 seconds 

PR Only intended for discussion. Not planning to land.

### CODE REVIEW

**Comment by andll** [comment]
I think its good in terms of push implementation, but on API side should we unify it to have same API as other coutners? @davidiw are you ok with pulling prometheus dependency?

File: consensus/safety-rules/src/lib.rs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 #![forbid(unsafe_code)]
 
 mod consensus_state;
+pub mod counters;
```
Comment (davidiw) [inline consensus/safety-rules/src/lib.rs]: why pub?

File: consensus/safety-rules/src/simple_push_metrics.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                    metric.counter_name,
+                    metric.counter.load(Ordering::SeqCst)
+                );
+                data.push_str(&metric_data);
+            }
+            let response = ureq::post("http://pushgateway.example.org:9091/metrics/job/safety_rules")
+              .timeout_connect(10_000)
+              .send_string(&data);
+            if response.ok() {
+                println!("Ok")
```
Comment (davidiw) [inline consensus/safety-rules/src/simple_push_metrics.rs]: should use logging crate debug / trace

File: consensus/safety-rules/src/simple_push_metrics.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            } else {
+                println!("Error")
```
Comment (davidiw) [inline consensus/safety-rules/src/simple_push_metrics.rs]: should use logging crate warn / error

File: consensus/safety-rules/src/simple_push_metrics.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+pub trait Metrics {
+    fn get_metrics(&self) -> Vec<&Counter>;
+}
+
+pub struct CountersPusher {}
+
+impl CountersPusher {
+    pub fn start(&self, metrics: Arc<dyn Metrics + Send + Sync>) -> JoinHandle<()> {
+        thread::spawn(move || loop {
+            let mut data = "".to_string();
```
Comment (davidiw) [inline consensus/safety-rules/src/simple_push_metrics.rs]: ideally this is broken into function calls so we can perform some form of integration tests

**Comment by davidiw** [comment]
> I think its good in terms of push implementation, but on API side should we unify it to have same API as other coutners? @davidiw are you ok with pulling prometheus dependency?

Here's cargo geiger on prometheus:
```
Metric output format: x/y
    x = unsafe code used by the build
    y = total unsafe code found in the crate

Symbols:
    🔒  = No `unsafe` usage found, declares #![forbid(unsafe_code)]
    ❓  = No `unsafe` usage found, missing #![forbid(unsafe_code)]
    ☢️   = `unsafe` usage found

Functions  Expressions  Impls  Traits  Methods  Dependency

0/0        339/344      0/0    0/0     0/0      ☢️  prometheus 0.8.0
0/0        0/0          0/0    0/0     0/0      ❓ ├── cfg-if 0.1.10
0/0        0/0          0/0    0/0     0/0      ❓ ├── fnv 1.0.6
0/0        7/7          1/1    0/0     0/0      ☢️  ├── lazy_static 1.4.0
2/2        2433/2436    0/0    0/0     1/1      ☢️  ├── protobuf 2.10.1
0/0        49/49        6/6    0/0     3/3      ☢️  ├── spin 0.5.2
0/0        0/0          0/0    0/0     0/0      ❓ └── thiserror 1.0.9
0/0        0/0          0/0    0/0     0/0      ❓     └── thiserror-impl 1.0.9
0/0        0/0          0/0    0/0     0/0      ❓         ├── proc-macro2 1.0.9
0/0        0/0          0/0    0/0     0/0      ❓         │   └── unicode-xid 0.2.0
0/0        0/0          0/0    0/0     0/0      ❓         ├── quote 1.0.2
0/0        0/0          0/0    0/0     0/0      ❓         │   └── proc-macro2 1.0.9
0/0        42/42        3/3    0/0     2/2      ☢️          └── syn 1.0.11
0/0        0/0          0/0    0/0     0/0      ❓             ├── proc-macro2 1.0.9
0/0        0/0          0/0    0/0     0/0      ❓             ├── quote 1.0.2
0/0        0/0          0/0    0/0     0/0      ❓             └── unicode-xid 0.2.0
```

There's quite a bit of unsafe code, but let's assume we can live with that, do we rely on the protobuf dependency?

```
[dependencies.protobuf]
version = "2.0"
optional = true
```

We can disable it if we are using our own interface.

**Comment by ankushagarwal** [comment]
> If we want greater granularity or attempted granularity of data, we could implement the counters to include a channel and send an event to the thread to push latest metrics whenever we update the metric. That way we can eliminate the loop and try to retain as high resolution as possible.

I am not convinced this is a good idea for a few reasons:

* Metrics which update at a very high frequency can cause high frequency updates to the push gateway server (or we would have to implement some sort of batching of requests)
* Metrics which don't get updated frequently - (eg: errors logged) wouldn't get pushed frequently (or at all) and can introduce confusion when trying to graph them.
* Since we are just a single server sending less than a dozen metrics, we can send metrics with as much (reasonable) granularity as possible. We probably don't want to send metrics with granularity less than a second - because I am not sure if it adds value (let me know if you think otherwise). Structured logging is probably a better place for tracking every event.
* Mostly it is against prometheus best practices to have different metrics to be collected at different timestamps from the same server

Re: most code comments - this was mostly a 30 min Proof of Concept on how simple it is to implement a basic push metrics system without prometheus dependency. 

I'll probably rewrite it using regular coding conventions if we can agree on the general direction of this project.

> I think its good in terms of push implementation, but on API side should we unify it to have same API as other coutners? @davidiw are you ok with pulling prometheus dependency?but on API side should we unify it to have same API as other coutners

* We can mould our API to look similar to Prometheus counter definition API
* Given that this is safety rules dependency, we should have a very strong reason for taking any new dependencies and I don't think this is a strong enough reason.

**Comment by davidiw** [comment]
> > If we want greater granularity or attempted granularity of data, we could implement the counters to include a channel and send an event to the thread to push latest metrics whenever we update the metric. That way we can eliminate the loop and try to retain as high resolution as possible.
> 
> I am not convinced this is a good idea for a few reasons:
> 
> * Metrics which update at a very high frequency can cause high frequency updates to the push gateway server (or we would have to implement some sort of batching of requests)
> * Metrics which don't get updated frequently - (eg: errors logged) wouldn't get pushed frequently (or at all) and can introduce confusion when trying to graph them because of all the missing data points.
> * Since we are just a single server sending less than a dozen metrics, we can send metrics with as much (reasonable) granularity as possible. We probably don't want to send metrics with granularity less than a second - because I am not sure if it adds value (let me know if you think otherwise). Structured logging is probably a better place for tracking every event.
> * Mostly it is against prometheus best practices to have different metrics to be collected at different timestamps from the same server
> 

I see, I think there might be some space to iterate here, so we have as fresh possible information, but I guess if prometheus is already pulling from other services, it is likely already losing that granularity ... is it? If it isn't then I think we could explore an ondemand push. I am not a big fan of pushing a metric just to feign liveness. All it tells you is that your process is alive and the metric thread is working but nothing else. We can likely get that information other ways. FWIW, I don't think any of what I said belongs in an MVP.

**Comment by ankushagarwal** [comment]
> if prometheus is already pulling from other services, it is likely already losing that granularity

yes, we pull only once every 15 seconds from all servers today and don't have granularity less than that.

> FWIW, I don't think any of what I said belongs in an MVP.

Makes sense. For the MVP, the bare minimum features that I want to support are:

1. ability to define gauge and counter metrics
2. ability to increment/set these counters in code
3. push metrics periodically (all at once) with a configurable period to a configurable server 

This set of features would take a fairly small amount of development effort because of the simple model and would get very close to the metrics that we have in libra core

After we have a solid MVP, we can try to implement other features based on how valuable they are
1. supporting histogram like metrics
2. pushing different metrics with different frequency (or on-demand)

**Comment by andll** [comment]
@ankushagarwal what is your vision in terms of unification with other crates?
All the features that you mentioned already exists in prometheus counter we use across codebase.

Is your vision to replace all the counters with our own implementation as a next step?

Or is TCB going to have similarly looking but different counters?

**Comment by ankushagarwal** [comment]
> @ankushagarwal what is your vision in terms of unification with other crates?
> All the features that you mentioned already exists in prometheus counter we use across codebase.

The way i see it is we will have two separate crates. `libra-metrics` built on prometheus cannot be used by TCB because of the unnecessary deps it pulls in. We can build a `libra-metrics-secure` which does not have any additional dependencies from what TCB already has and is built on the push model. 

> Is your vision to replace all the counters with our own implementation as a next step?
> Or is TCB going to have similarly looking but different counters?

No, I don't plan on doing that. Non-TCB codebase can continue using `libra-metrics` and TCB codebase will only use `libra-metrics-secure` with its own counters.

**Comment by andll** [comment]
I think we should avoid duplicated solution.
For @davidiw's comment above I understand that he was mostly ok with prometheus dependency, if we remove optional protobuf dependency from it. If we can keep prometheus, we should preserve as much as we can, instead of doing something completely separate.

**Pull request closed without merging**